### PR TITLE
Overhauling solver especially boundary conditions

### PR DESCRIPTION
**Pull request by sebglane**

### CODE REVIEW

File: demo/cavity_flow.py [hunk 1]
```diff
@@ -30,12 +30,10 @@ def set_boundary_conditions(self):
...
-        velocity_bcs = (
-                (no_slip, BoundaryMarkers.left.value, None),
-                (no_slip, BoundaryMarkers.right.value, None),
-                (no_slip, BoundaryMarkers.bottom.value, None),
-                (constant, BoundaryMarkers.top.value, (1.0, 0.0)))
-        self._bcs = {"velocity": velocity_bcs}
+        self._bcs = ((no_slip, BoundaryMarkers.left.value, None),
+                     (no_slip, BoundaryMarkers.right.value, None),
+                     (no_slip, BoundaryMarkers.bottom.value, None),
+                     (constant, BoundaryMarkers.top.value, (1.0, 0.0)))
```
Comment (liuyiming0507) [inline demo/cavity_flow.py:36]: Why we have these changes about the definition of boundary conditions

Comment (idanven) [reply demo/cavity_flow.py:36]: just want to omit this line?  ` self._bcs = {"velocity": velocity_bcs}`

Comment (sebglane) [reply demo/cavity_flow.py:36]: The reason for this change is due to changes in the underlying problem and solver classes. The way the boundary conditions are specified has changed in this branch.

Comment (sebglane) [reply demo/cavity_flow.py:36]: Moreover, it is clear from
```
(VelocityBCType.no_slip, BoundaryMarkers.left.value, None)
```
that this is a velocity boundary condition. Therefore, the mapping is not really required.

**Pull request merged**